src/Microsoft.Health.Dicom.Core/Features/Store/StoreDatasetValidator.cs

```diff
@@ -194,36 +197,64 @@ string EnsureRequiredTagIsPresent(DicomTag dicomTag)
             }
             catch (DicomValidationException ex)
             {
-                if (EnableDropMetadata(_dicomRequestContextAccessor.RequestContext.Version))
+                validationResultBuilder.Add(ex, item.Tag);
+            }
+        }
+    }
+
+    private async Task ValidateAllItemsWithLeniencyAsync(
+        DicomDataset dicomDataset,
+        StoreValidationResultBuilder validationResultBuilder)
+    {
+        IReadOnlyCollection<QueryTag> queryTags = await _queryTagService.GetQueryTagsAsync();
+        foreach (DicomItem item in dicomDataset)
+        {
+            try
+            {
+                string value = dicomDataset.GetString(item.Tag);
+                if (value.EndsWith('\0'))
                 {
-                    if (RequiredCoreTags.Contains(item.Tag))
-                    {
-                        validationResultBuilder.Add(ex, item.Tag, isCoreTag: true);
-                    }
-                    else
-                    {
-                        validationResultBuilder.Add(ex, item.Tag);
-                    }
-
-                    bool isIndexableTag = queryTags.Any(x => x.Tag == item.Tag);
-                    _storeMeter.ValidateAllValidationError.Add(
-                        1,
-                        new[]
-                        {
-                            new KeyValuePair<string, object>("TagKeyword", item.Tag.DictionaryEntry.Keyword),
-                            new KeyValuePair<string, object>("VR", item.ValueRepresentation.ToString()),
-                            new KeyValuePair<string, object>("Tag", item.Tag.ToString()),
-                            new KeyValuePair<string, object>("IsIndexable", isIndexableTag.ToString())
-                        });
+                    ValidateWithoutNullPadding(value, item, queryTags);
                 }
                 else
                 {
-                    validationResultBuilder.Add(ex, item.Tag);
+                    item.Validate();
                 }
             }
+            catch (DicomValidationException ex)
+            {
+                validationResultBuilder.Add(ex, item.Tag, isCoreTag: RequiredCoreTags.Contains(item.Tag));
+
+                bool isIndexableTag = queryTags.Any(x => x.Tag == item.Tag);
+                _storeMeter.V2ValidationError.Add(
+                    1,
+                    TelemetryDimension(item, isIndexableTag));
+            }
         }
     }
 
+    private void ValidateWithoutNullPadding(string value, DicomItem item, IReadOnlyCollection<QueryTag> queryTags)
+    {
+        // we still want to save the original value when validation passes so create new ds to validate with
+        var newValue = value.TrimEnd('\0');
+        DicomDataset ds = new DicomDataset().NotValidated();
```

Comment:
Do you need to create the DicomDataset object here? Looks like you just need DicomItem with the tag and trimmed value to do the validation.

Reply:
I agree, I would not go to reflection, How about this, change your for - loop

//Change to DicomElement, it is better object DicomItem just as tag.
foreach (DicomElement item in dicomDataset)
{

try
{
IgnoredNullPaddingValidation(DicomElement)
}
catch
{
}
}

Now change your method
private void IgnoredNullPaddingValidation(DicomElement de)
{
       var stringValue = de.Get<string>(-1);
    
       if (stringValue.EndsWith('\0'))
       {
           //log
           de.ValueRepresentation.ValidateString(stringValue.TrimEnd('\0'));
       }
       else
       {
             de.Validate()
        }
}